File: puccini-tosca/cmd/compile.go
package cmd

import (
	"github.com/spf13/cobra"
	"github.com/tliron/puccini/common"
	format_ "github.com/tliron/puccini/common/format"
	"github.com/tliron/puccini/common/terminal"
	"github.com/tliron/puccini/tosca/compiler"
)

var output string
var resolve bool
var coerce bool

func init() {
	rootCmd.AddCommand(compileCmd)
	compileCmd.Flags().StringArrayVarP(&inputs, "input", "i", []string{}, "specify an input (name=YAML)")
	compileCmd.Flags().StringVarP(&inputsUrl, "inputs", "n", "", "load inputs from a PATH or URL")
	compileCmd.Flags().StringVarP(&output, "output", "o", "", "output Clout to file (default is stdout)")
	compileCmd.Flags().BoolVarP(&resolve, "resolve", "r", true, "resolves the topology (attempts to satisfy all requirements with capabilities)")
	compileCmd.Flags().BoolVarP(&coerce, "coerce", "c", false, "coerces all values (calls functions and applies constraints)")
}

var compileCmd = &cobra.Command{
	Use:   "compile [[TOSCA PATH or URL]]",
	Short: "Compile TOSCA to Clout",
	Long:  `Parses a TOSCA service template and compiles the normalized output of the parser to Clout. Supports JavaScript plugins.`,
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		var url string
		if len(args) == 1 {
			url = args[0]
		}

		Compile(url)
	},
}

func Compile(url string) {
	// Parse
	context, s := Parse(url)
	problems := context.GetProblems()

	// Compile
	clout, err := compiler.Compile(s)
	common.FailOnError(err)

	// Resolve
	if resolve {
		compiler.Resolve(clout, problems, format, pretty)
		FailOnProblems(problems)
	}

	// Coerce
	if coerce {
		compiler.Coerce(clout, problems, format, pretty)
		FailOnProblems(problems)
	}

	if !terminal.Quiet || (output != "") {
		err = format_.WriteOrPrint(clout, format, terminal.Stdout, pretty, output)
		common.FailOnError(err)
	}
}

File: examples/ruby/compile.rb
#!/usr/bin/env ruby

require 'optparse'
require 'puccini'
require 'yaml'

inputs = {}
quirks = []
resolve = true
coerce = true

option_parser = OptionParser.new do |opts|
  opts.on '-i', '--input=INPUT', 'specify input (format is name=value)' do |i|
    k, v = i.split('=')
    inputs[k] = YAML.load(v)
  end
  opts.on '-q', '--quirk=QUIRK', 'specify quirk' do |q|
    quirks.push(q)
  end
  opts.on '-r', '--resolve=FLAG', 'whether to resolve' do |r|
    resolve = (r == 'true')
  end
  opts.on '-c', '--coerce=FLAG', 'whether to coerce' do |c|
    coerce = (c == 'true')
  end
end

option_parser.parse!

if ARGV.length == 0
  puts 'no URL provided'
  exit 1
end

begin
  clout = Puccini::TOSCA.compile(ARGV[0], inputs, quirks, resolve, coerce)
  puts YAML.dump(clout)
rescue Puccini::TOSCA::Problems => e
  puts 'Problems:'
  for problem in e.problems
    puts YAML.dump(problem)
  end
  exit 1
end
